### PR TITLE
[fix]: refactor code to replace time.Now() with Now() in clock package in pkg/scheduler(#3380)

### LINKED ISSUE #3380: Replace `time.Now()` code with `clock.Clock.Now()`
**What would you like to be added**:
There are packages like e.g. `pkg/scheduler` that use `time.Now()`. As a result this code is hard to test and prone to flakiness as we are not able to directly control the time during tests. It should be replaced with `clock.Clock.Now()` as it is  in WorkloadController

**Why is this needed**:

**Completion requirements**:

This enhancement requires the following artifacts:

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by TusharMohapatra07**
#### What type of PR is this?
/kind feature
/kind flake

#### What this PR does / why we need it:
Replaces `time.Now()` to use the Now() function present inside `clock` package

#### Which issue(s) this PR fixes:
Fixes #3380 

#### Special notes for your reviewer:
This PR is work-in-progress. I've refactored the inside `pkg/scheduler` for now. I shall proceed to refactor more post review.

```release-note
NONE
```

### CODE REVIEW

**Comment by mimowo** [comment]
cc @mbobrovskyi for first pass

**Review by mbobrovskyi — COMMENTED** [review]
- Don't forget to change `time.Since(...)` to `s.clock.Since(...)` as well.
- Please update scheduler_test.go using `FakeClock` instead of time.
- Implement new `Option` to provide `FakeClock` to scheduler.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi can you please elaborate on the `Option` thing ? i couldn't exactly understand the requirement. Also if there's any way i could join any slack or discord , that would be great as it would make a direct source of communication. I tried to join the slack channel but it's showing that it couldn't find my gmail.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi can you please elaborate on the Option thing ?

We already have a few options on Scheduler
https://github.com/kubernetes-sigs/kueue/blob/55e136718962bbc5e7d52a134ca4ee32657f60dd/pkg/scheduler/scheduler.go#L98-L110

So you need just to add one more like here 
https://github.com/kubernetes-sigs/kueue/blob/55e136718962bbc5e7d52a134ca4ee32657f60dd/pkg/controller/jobframework/reconciler.go#L194-L201

> Also if there's any way i could join any slack or discord , that would be great as it would make a direct source of communication. I tried to join the slack channel but it's showing that it couldn't find my gmail.

Yes, here https://kubernetes.slack.com/.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi 
image
i get this mail when i try to login.

**Comment by mbobrovskyi** [comment]
Try this https://slack.k8s.io/

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi please check if the 1st and 3rd requirement are being fulfilled.

about the 2nd one, is it similar to what's implemented here https://github.com/kubernetes-sigs/kueue/blob/13e77053fcef3e2c9202112f41ff83e51d58e4c0/pkg/controller/core/workload_controller_test.go#L48 ?

**Comment by mbobrovskyi** [comment]
> kueue/pkg/controller/core/workload_controller_test.go

Yes, exactly.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi so far i've figured it out that `Fakeclock` implements clock.Clock and has additional methods but i couldn't find any potential places where it could be replaced with something. Please help

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi so far i've figured it out that `FakeClock` implements clock.Clock and has additional methods but i couldn't find any potential places where it could be replaced with something. Please help

I think here 

https://github.com/kubernetes-sigs/kueue/blob/ddc235444a485093f99e8a8ab74fd481a82878b5/pkg/scheduler/scheduler_test.go#L94

You need to add 

https://github.com/kubernetes-sigs/kueue/blob/13e77053fcef3e2c9202112f41ff83e51d58e4c0/pkg/controller/core/workload_controller_test.go#L48

And on the 

https://github.com/kubernetes-sigs/kueue/blob/ddc235444a485093f99e8a8ab74fd481a82878b5/pkg/scheduler/scheduler_test.go#L2654

use already created Option.

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -126,6 +134,7 @@ func New(queues *queue.Manager, cache *cache.Cache, cl client.Client, recorder r
 		preemptor:               preemption.New(cl, wo, recorder, options.fairSharing, realClock),
 		admissionRoutineWrapper: routine.DefaultWrapper,
 		workloadOrdering:        wo,
+		clock:                   realClock,
```
Comment (mbobrovskyi) [inline pkg/scheduler/scheduler.go:137]: You steal using realClock. Use clock from options.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -126,6 +134,7 @@ func New(queues *queue.Manager, cache *cache.Cache, cl client.Client, recorder r
 		preemptor:               preemption.New(cl, wo, recorder, options.fairSharing, realClock),
```
Comment (mbobrovskyi) [inline pkg/scheduler/scheduler.go:134]: And also here.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -109,6 +111,12 @@ func WithFairSharing(fs *config.FairSharing) Option {
...
 }
 
+func WithClock(_ testing.TB, c clock.Clock) Option {
+	return func(o *options) {
+		o.clock = c
+	}
+}
+
 func New(queues *queue.Manager, cache *cache.Cache, cl client.Client, recorder record.EventRecorder, opts ...Option) *Scheduler {
 	options := defaultOptions
```
Comment (mbobrovskyi) [inline pkg/scheduler/scheduler.go:121]: Don't forget to add realClock to defaultOptions.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi so primarily, `FakeClock` used to initialise the clock field in a `Scheduler` instance. Am i right ?

By default it should be initialised by realClock. On the test we need to replace it with FakeClock.

**Comment by TusharMohapatra07** [comment]
understood

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (TusharMohapatra07) [reply pkg/scheduler/scheduler.go:137]: I didn't understand.... Please elaborate

File: pkg/scheduler/scheduler_test.go [hunk 4]
```diff
@@ -92,6 +93,8 @@ const (
 
 func TestSchedule(t *testing.T) {
 	now := time.Now()
+	fakeClock := testingclock.NewFakeClock(now)
```
Comment (mbobrovskyi) [inline pkg/scheduler/scheduler_test.go:96]: Please also do the same in other tests on this file.

**Review by mbobrovskyi — COMMENTED** [review]
Almost done. :)

Also please fix test with `make lint-fix`.

**Comment by TusharMohapatra07** [comment]
hey @mbobrovskyi, please review and suggest changes if any

**Comment by mbobrovskyi** [comment]
> hey @mbobrovskyi, please review and suggest changes if any

Please also cover this case 
https://github.com/kubernetes-sigs/kueue/blob/f7d342b63c236ea2f18eadd30afde2cbf09993b0/pkg/scheduler/scheduler_test.go#L2765

**Review by mbobrovskyi — COMMENTED** [review]
Good job! Thanks! 🙏

/lgtm

**Comment by mbobrovskyi** [comment]
/release-note-edit

```release-note
NONE
```

**Comment by mbobrovskyi** [comment]
@TusharMohapatra07 Please add 

\`\`\`release-note
NONE
\`\`\`

on the PR description

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi about the case that you asked me to cover https://github.com/kubernetes-sigs/kueue/blob/f7d342b63c236ea2f18eadd30afde2cbf09993b0/pkg/scheduler/scheduler_test.go#L2765 i couldn't find any scheduler instance being created inside the test function to pass the fakeClock to

**Comment by mbobrovskyi** [comment]
@TusharMohapatra07 Also please remove WIP on PR title if it's ready to review.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi what about other packages ? should i raise another pr for those ? i think it would be a good idea to make one PR for each package refactor

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi what about other packages ? should i raise another pr for those ? i think it would be a good idea to make one PR for each package refactor

Yes, feel free to follow up on separate PRs.

**Comment by TusharMohapatra07** [comment]
@mbobrovskyi done

**Comment by mimowo** [comment]
/approve
thanks folks!

**Pull request merged**